# Evolution_of_gene_expression_in_seasonal_envrionment
This repository contains R scripts for the paper "Evolution of gene expression in seasonal environments." The scripts in this repository contains processing of gene expression data from raw transcript counts and comparative analyses of seasonal gene expression across four dominant evergreen Fagaceae species (*Quercus glauca*, *Q. acuta*, *Lithocarpus edulis*, and *L. glaber*). To investigate the evolutioin of gene expression patterns under seasonal environments, this repository also include the scripts for identification of genes with seasonal rhythmic expressions, quantification of their periods and peak months, and comparison of these metrics across species.

# Overview
## 1. Data processing
This script processes gene expression data through multiple steps to prepare it for comparative analysis. The steps include:
- Filtering low-expression genes  
  We excluded the genes with low expression level using the threshold, the value of mean Reads Per Kilobase (RPK) < 1 for each species.
- Converting count data to GeTMM values  
  We normalized the expression level using the value of Gene length corrected Trimmed Mean of M-values (GeTMM) using an R package "edgeR".

- Extracting single-copy orthologous genes  
To compare the seasonal expression pattern of each gene across species, we identified single-copy orthologs based on the Orthogroup information estimated by Orthofinder2 after filtering out genes mapped outside the 12 chromosomes.

**Input**: count files generated by RSEM (concatenated for each species), list of single-copy orthologs  
**Output**: a matrix of gene expression data whose rows have gene IDs and columns hold sample IDs.


## 2. Hierarchical Clustering
This script performs hierarchical clustering on both samples and genes based on gene expression to assess the similarity of gene expression profiles across tissues, speices, and seasons. Additionally, it evaluates the relationship between sample clustering and climatic conditions. The results are used to generate Figure 2 in the manuscript.  

**Input**: gene expression data matrix generated by the script above, climatic information (daily-mean temperature and photoperiod of each study sites on each sampling date).  
**Output**: heatmap showing clustering results, samle and gene list with cluster information.


## 3. Principal Component Analysis (PCA)
This script performs Principal Component Analysis (PCA) to reduce the dimension of gene expression data and assess the similarity of gene expression profiles across tissues, speices, and seasons. The results are used to generate Figure S6 in the manuscript.  

**Input**: gene expression data matrix, climatic information (daily-mean temperature of each study sites on each sampling date).  
**Output**: 3D plot showing the first principal component (PC1), second (PC2), and third principal component (PC3).



## 4. Rhythm Detection
This script detects rhythmic gene expression patterns using the RAIN algorithm (Thaben and Westermark, 2014) and calculates the period and peak in seasonal gene expression. Based on these information, we first classified genes into "rhythmic" and "arrhythmic" using Benjamini-Hochberg (BH) adjusted q-value threshold of < 0.01. Then, we further classified rhythmic genes into two categories based on the period length T (months): "half-annual" (0<=T<8), "annual" (8 <= T <= 16). Genes with periods longer than 16 months were classified as arrhythmic because of the low confidence.

**Input**: gene expression data matrix.  
**Output**: gene list with rhythmicity information estimated by RAIN.


## 5. Species Comparison
This script performs a comparative analysis of gene expression patterns across species. To quantify the similarity of seasonal expression pattern of each gene between each pair of species, we calculated following metrics:

- Pearson's correlation  
- Mean absolute difference
- Seasonal peak difference

Furthermore, to examine the phylogenetic constraints on the seasonal gene expression evolution, we compared the distribution of Pearson's correlation across the species pairs using Nemenyi test.

In addition, we performed Gene Ontology (GO) enrichment analysis of genes with conserved/diverged expression pattern extracted based on Pearson's correlation.

**Input**: gene expression data matrix, list of corresponding time points to align seasonal cycles of each pair of species, rhythmicity information estimated by RAIN.  
**Output**: gene list with each metrics, result of GO analysis.

## 6. Evolutionary Rate Analysis
This script examines the relationship between gene expression patterns and evolutionary rates. It requires the information of evolutionary measures (dN, dS, and dN/dS) estimated in advance. It involves:

- Outlier filtering  
We excluded the genes with extremely large or small evolutionary rates based on the distribution of dN and dS across genes.

- Linear regression analysis (peak difference, mean expression levels vs dN/dS)  
Linear regression analyses were performed to examine the relationship between dN, dS, dN/dS and peak month differences (Δφ) in gene expression between Q. glauca and L. edulis, as well as  mean expression levels.  
We modeled the evolutionary rates (dN, dS, and dN/dS) as the response variable and the peak month difference (Δφ) or the mean levels in gene expression as the predictor variable.

- Multiple comparisons (Steel-Dwass test) of dN/dS
To examine the seasonal difference of sequence divergence, We classified genes that exhibited annual periodicity across both species and tissues into four groups based on their expression peak timing (winter or growing seasons) and tissue specificity (leaves or buds) and compared the extent of protein sequence divergence across these groups.

**Input**: gene expression data matrix, rhythmicity information estimated by RAIN, gene expression similarity metrics calculated in section 5, and list of evolutionary rates across genes.  
**Output**: Statistical significance of the linear regression and the multiple comparison of evolutionary rates.


# Usage
Each script is designed to be run in R. Ensure that all required packages are installed before executing the scripts. Detailed instructions on data input formats and dependencies are provided within each script.

